### PR TITLE
Added support for floating point operators

### LINKED ISSUE #38: Support floating point operators
Currently operator mutation only supports integer operators. It would be good to support floating point types too.

### PR DESCRIPTION
**Pull request by JamesLee-Jones**
Fixes #38

### CODE REVIEW

File: src/libdredd/src/mutate_visitor.cc
Comment (afd) [inline src/libdredd/src/mutate_visitor.cc:103]: Now that this check has gotten a bit more complicated, I think there's a case for extracting out the logic into a private helper method, `IsTypeSupported`.

This should be prototyped in the class, in `mutate_visitor.h`, and implemented in `mutate_visitor.cc`.

File: src/libdredd/src/mutate_visitor.cc
Comment (afd) [inline src/libdredd/src/mutate_visitor.cc:109]: Call the helper.

File: src/libdredd/src/mutate_visitor.cc
Comment (afd) [inline src/libdredd/src/mutate_visitor.cc:115]: Call the helper.

**Review by afd — CHANGES_REQUESTED** [review]
The code changes look good! One comment about refactoring.

However, this change needs some tests.

Can you please:

- Pull main and rebase over my latest changes
- Add a couple of tests to `mutate_binary_operator_test.cc`, along the lines of the ones that are there, but for floating-point operators; e.g. `/` on `float` and `-=` on `double` (rather arbitrarily)
- Add a small synthetic example that does a few floating-point operations under `test/single_file`, with a corresponding expectation file. See the README under `test` for how to generate the expectation file automatically - but do check that it looks right before you commit it.

File: src/libdredd/src/mutate_visitor.cc
Comment (JamesLee-Jones) [reply src/libdredd/src/mutate_visitor.cc]: Done

File: src/libdredd/src/mutate_visitor.cc
Comment (JamesLee-Jones) [reply src/libdredd/src/mutate_visitor.cc]: Done

File: src/libdredd/src/mutate_visitor.cc
Comment (JamesLee-Jones) [reply src/libdredd/src/mutate_visitor.cc]: Done

File: src/libdreddtest/src/mutation_replace_binary_operator_test.cc
Comment (afd) [reply src/libdreddtest/src/mutation_replace_binary_operator_test.cc:262]: Similar comment for other uses of num_replacements above.

File: src/libdredd/src/mutate_visitor.cc
Comment (afd) [inline src/libdredd/src/mutate_visitor.cc:106]: Given that `result_type`, `lhs_type` and `rhs_type` are never used again, it would be better to change `IsTypeSupported` so that it accepts a `QualType` as an argument, and then you can just do:

`if (IsTypeSupported(binary_operator->getType())`

and similar below with `getType()` on the LHS and RHS.

The `getAs<clang::BuiltinType>()` can then go inside the `IsTypeSupported` method.

**Review by afd — CHANGES_REQUESTED** [review]
This is getting there - see latest batch of comments.

File: src/libdredd/src/mutate_visitor.cc
Comment (afd) [inline src/libdredd/src/mutate_visitor.cc:41]: nit: I would call the variable `builtin_type`, to better distinguish it from `qual_type`.

File: src/libdredd/src/mutation_replace_binary_operator.cc
Comment (afd) [inline src/libdredd/src/mutation_replace_binary_operator.cc:59]: I would:

- add assertions that both lhs_type and rhs_type are either integer or floating-point (this should be true by construction, but it's good to confirm); use the && trick to have a suitable string associated with the assertions.
- add a comment noting that certain binary operators are not compatible with floating-point arguments.

**Review by afd — CHANGES_REQUESTED** [review]
This looks great! I have noted a couple of small things, and would normally approve and let you do them and then merge. However, since you're going to have to rework your changes slightly once I land my pending PR, I've hit "request changes" so that I can take a quick look at your revision in due course.

**Comment by afd** [comment]
@JamesLee-Jones Ah, the CI already fails because I made a change yesterday that uses if statements instead of switch statements when checking which binary operator mutation to apply. I suggest you hold off until I land my pending change about boolean arrays, and then you can fix things up.

File: src/libdredd/src/mutate_visitor.cc
Comment (JamesLee-Jones) [reply src/libdredd/src/mutate_visitor.cc]: Done.

File: src/libdredd/src/mutation_replace_binary_operator.cc
Comment (JamesLee-Jones) [reply src/libdredd/src/mutation_replace_binary_operator.cc:65]: Done.

**Review by afd — APPROVED** [review]
Looks great!

**Pull request merged**
